### PR TITLE
correct free resources on exceptions

### PR DESCRIPTION
**Pull request by kostya**

### CODE REVIEW

File: src/icu/charset_detector.cr [hunk 1]
```diff
@@ -26,13 +26,21 @@ class ICU::CharsetDetector
   @@detectable_charsets : Array(String)?
 
   def initialize
+    @finalized = false
     ustatus = LibICU::UErrorCode::UZeroError
     @csdet = LibICU.ucsdet_open(pointerof(ustatus))
-    ICU.check_error!(ustatus)
+    ICU.check_error!(ustatus) { free }
```
Comment (olbat) [inline src/icu/charset_detector.cr:32]: I'm not sure free should be called here, [the doc](http://userguide.icu-project.org/dev/codingguidelines#TOC-Functions-for-Constructors-and-Destructors) specifies:

> Each successful call to a umod_open() returns a pointer to an object that must be released by the user/owner by calling the matching umod_close().

So for me if the call is not successful, nothing should be freed.

**Comment by olbat** [comment]
I don't know Crystal enough to be 100% sure but aren't the resources supposed to be freed by the GC even if there is a non-catched exception ? (+ I'm not sure that objects resulting from failed calls to `*_open` should be freed, cf. the review)

**Comment by kostya** [comment]
this function need in still needed in multiple allocations like in break_iterator

**Comment by olbat** [comment]
Erm, which one ? (if I did not miss anything, the one that are modified in your commit only includes a single allocation)

**Pull request closed without merging**